interpreter/src/cursor.rs

```diff
@@ -0,0 +1,3003 @@
+};
+
+use indexmap::{IndexMap, IndexSet};
+
+use rand::Rng as _;
+
+use rand_chacha::{ChaCha20Rng, rand_core::SeedableRng};
+
+use snarkvm::prelude::{
+    Address as SvmAddressParam,
+    Boolean as SvmBooleanParam,
+    Cast,
+    CastLossy as _,
+    Double as _,
+    Field as SvmFieldParam,
+    FromBits as _,
+    Group as SvmGroupParam,
+    Identifier as SvmIdentifierParam,
+    Inverse as _,
+    Network as _,
+    Pow as _,
+    Scalar as SvmScalarParam,
+    // Signature as SvmSignatureParam,
+    SizeInBits,
+    Square as _,
+    SquareRoot as _,
+    TestnetV0,
+    ToBits,
+    integers::Integer as SvmIntegerParam,
+};
+
+type SvmAddress = SvmAddressParam<TestnetV0>;
+type SvmBoolean = SvmBooleanParam<TestnetV0>;
+type SvmField = SvmFieldParam<TestnetV0>;
+type SvmGroup = SvmGroupParam<TestnetV0>;
+type SvmIdentifier = SvmIdentifierParam<TestnetV0>;
+type SvmInteger<I> = SvmIntegerParam<TestnetV0, I>;
+type SvmScalar = SvmScalarParam<TestnetV0>;
+// type SvmSignature = SvmSignatureParam<TestnetV0>;
+
+use leo_ast::{
+    AccessExpression,
+    AssertVariant,
+    BinaryOperation,
+    Block,
+    CoreConstant,
+    CoreFunction,
+    Expression,
+    Function,
+    GroupLiteral,
+    IntegerType,
+    Literal,
+    Node as _,
+    Statement,
+    Type,
+    UnaryOperation,
+    Variant,
+};
+
+use leo_span::{Span, Symbol};
+
+use leo_errors::{InterpreterHalt, Result};
+
+#[macro_export]
+macro_rules! tc_fail {
+    () => {
+        panic!("type checker failure")
+    };
+}
+
+macro_rules! halt {
+    ($span: expr) => {
+        return Err(InterpreterHalt::new_spanned(String::new(), $span).into())
+
+    };
+
+    ($span: expr, $($x:tt)*) => {
+        return Err(InterpreterHalt::new_spanned(format!($($x)*), $span).into())
+    };
+}
+
+trait ExpectTc {
+    type T;
+    fn expect_tc(self, span: Span) -> Result<Self::T>;
+}
+
+impl<T> ExpectTc for Option<T> {
+    type T = T;
+
+    fn expect_tc(self, span: Span) -> Result<Self::T> {
+        match self {
+            Some(t) => Ok(t),
+            None => Err(InterpreterHalt::new_spanned("type failure".into(), span).into()),
+        }
+    }
+}
+
+impl<T, U: std::fmt::Debug> ExpectTc for Result<T, U> {
+    type T = T;
+
+    fn expect_tc(self, span: Span) -> Result<Self::T> {
+        self.map_err(|_e| InterpreterHalt::new_spanned("type failure".into(), span).into())
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StructContents {
+    pub name: Symbol,
+    pub contents: IndexMap<Symbol, Value>,
+}
+
+impl Hash for StructContents {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        for (_symbol, value) in self.contents.iter() {
+            value.hash(state);
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct AsyncExecution {
+    pub function: GlobalId,
+    pub arguments: Vec<Value>,
+}
+
+#[derive(Clone, Debug, Default, Eq, PartialEq, Hash)]
+pub struct Future(pub Vec<AsyncExecution>);
```

Comment:
That way I've been thinking about `Future`s is that they are a single `AsyncExecution`, whose `arguments` can be other `AsyncExecution`s. I'm sure I'll understand as I make it further down, but is there a reason why it's a vec here?